synapse/handlers/oidc_handler.py

```diff
@@ -906,20 +908,34 @@ async def _map_userinfo_to_user(
         localpart = map_username_to_mxid_localpart(attributes["localpart"])
 
         user_id = UserID(localpart, self._hostname)
-        if await self._datastore.get_users_by_id_case_insensitive(user_id.to_string()):
-            # This mxid is taken
-            raise MappingException(
-                "mxid '{}' is already taken".format(user_id.to_string())
-            )
-
-        # It's the first time this user is logging in and the mapped mxid was
-        # not taken, register the user
-        registered_user_id = await self._registration_handler.register_user(
-            localpart=localpart,
-            default_display_name=attributes["display_name"],
-            user_agent_ips=(user_agent, ip_address),
+        users = await self._datastore.get_users_by_id_case_insensitive(
+            user_id.to_string()
         )
-
+        if users:
+            if self._allow_existing_users:
+                if len(users) == 1:
+                    registered_user_id = UserID.from_string(next(iter(users)))
+                elif user_id in users:
+                    registered_user_id = user_id
```

Comment:
`registered_user_id` eventually gets returned, so we want to ensure it is a `str` like the other places we return:

```suggestion
                    registered_user_id = next(iter(users))
                elif user_id in users:
                    registered_user_id = user_id.to_string()
```

Reply:
(I'm surprised mypy didn't complain about this...)

Reply:
Looks like this is because `get_user_by_external_id` is incorrectly typed and mypy is ignoring half the function.

Reply:
Maybe you'd like to include the fix in this PR?

```diff
diff --git a/synapse/storage/databases/main/registration.py b/synapse/storage/databases/main/registration.py
index 675e81fe3..c24b79435 100644
--- a/synapse/storage/databases/main/registration.py
+++ b/synapse/storage/databases/main/registration.py
@@ -379,7 +379,7 @@ class RegistrationWorkerStore(SQLBaseStore):
 
     async def get_user_by_external_id(
         self, auth_provider: str, external_id: str
-    ) -> str:
+    ) -> Optional[str]:
         """Look up a user by their external auth id
 
         Args:
@@ -387,7 +387,7 @@ class RegistrationWorkerStore(SQLBaseStore):
             external_id: id on that system
 
         Returns:
-            str|None: the mxid of the user, or None if they are not known
+            the mxid of the user, or None if they are not known
         """
         return await self.db_pool.simple_select_one_onecol(
             table="user_external_ids",
```

Reply:
Of course.